File: 5. LinkedLists/1. linked_list_values/linked_list_values.py
class Node:
    def __init__(self, value):
        self.val = value
        self.next = None

a = Node("A")
b = Node("B")
c = Node("C")
d = Node("D")

a.next = b
b.next = c
c.next = d

# def linked_list_values(head):
#   current = head
#   outputList = []
#   while(current is not None):
#     outputList.append(current.val)
#     current = current.next
#   return outputList

def print_list(head):
  output = ""
  while(head.next is not None):
    output += str(head.val) + "->"
    head = head.next
  output += str(head.val)
  print(output)

def linked_list_values(head):
  values = []
  _linked_list_values(head, values)
  return values
  
def _linked_list_values(head, values):
  if head is None:
    return
  values.append(head.val)
  _linked_list_values(head.next, values)

print(linked_list_values(a))
print_list(a)

File: 5. LinkedLists/1. linked_list_values/linked_list_values.js
// class Node {
//   constructor(val) {
//     this.val = val;
//     this.next = null;
//   }
// }

// const a = new Node("a");
// const b = new Node("b");
// const c = new Node("c");
// const d = new Node("d");

// a.next = b;
// b.next = c;
// c.next = d;



// Solution
// ---------
// Iterative
// const linkedListValues = (head) => {
//     arr = [];
//     while(head !== null){
//         arr.push(head.val);
//         head = head.next;
//     }
//     return arr;
// }

// Recursive - 1
// const linkedListValues = (head) => {
//   const values = [];
//   _linkedListValues(head, values);
//   return values;
// };

// const _linkedListValues = (head, values) => {
//   if (head === null) return;
//   values.push(head.val);
//   _linkedListValues(head.next, values);
// };

// Recursive - 2
// const linkedListValues = (head, finalArr = []) => {
//   if(head === null) return finalArr;
//   finalArr.push(head.val);
//   return linkedListValues(head.next, finalArr);
// }


// Recursive - 3
const linkedListValues = (head, arr=[]) => {
    if(head === null) return [];
    arr = [head.val, ...linkedListValues(head.next)]
    return arr;
}

// Complexity
// n = number of nodes
// Time: O(n)
// Space: O(n)

// let ans = linkedListValues(a);
// console.log(`Ans: ${ans}`);